### ISSUE TITLE
Malformed CIGAR string for non-primary aligment

### ISSUE THREAD

**Issue by flokraft85** [nanoporetech/dorado#496]
Hi dorado developer team,
I had some issues using dorado produced bam for downstream analysis with sniffles2. Using the bams as input for sniffles2, I got a lot of errors regarding to CIGAR string format.
Sniffles2 Error: Malformed CIGAR '31H2817M442I-732H' with pos 49709142 of read 'e4955234-38c2-425a-aef7-1d788738077f' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '-7614H4784M1443I4137H' with pos 43237603 of read 'e633dac1-2ecb-4ef3-9dfd-09517808b11f' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '6176H11730M4513I-8607H' with pos 105136 of read '6d3bf297-86f1-42e0-92be-2a03931fb064' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '98H4234M220D-2502H' with pos 11211033 of read '7579f2b1-8ae2-44c2-a310-a4ccc05c7c78' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '5616H13768M3410I-9749H' with pos 105136 of read '3ce4d068-f4aa-4224-9bf1-c6099d4f3a9d' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '1273H10032M878I-9702H' with pos 105136 of read '9b8e95f1-8d0d-4505-aa9f-138dce20676a' (invalid literal for int() with base 10: ''). Skipping.
Sniffles2 Error: Malformed CIGAR '4036H16234M3758I-11373H' with pos 105136 of read '2016fabb-acdd-4a0f-9811-1bd08ce22ae3' (invalid literal for int() with base 10: ''). Skipping.
(just a small excerpt of all errors).

When I use samtools to extract only primary alignments from the bam (samtools view -bF 0x900 -q 1 -@30 ./input.bam > ./primary.bam), sniffles2 worked without any issues. So i would guess the problem is only releated to the non-primary aligments.
I use the dorado version 0.4.2+bb2a133 for basecalling and alignment, with the following command.
dorado basecaller dna_r10.4.1_e8.2_400bps_hac\@v4.2.0 --modified-bases 5mCG_5hmCG --reference ref.mmi -b 5184 -r /input/ > /output.bam

When using sniffles2 after realignment of the bam with minimap2, I also had no issues. 
samtools fastq -@20 -t -T MM,ML ./input.bam | minimap2 -t 20 -ax map-ont  /ref.fa - | samtools sort -@ 20 -o ./output.bam

Thanks in advance investigating the issue!
Best,
Florian

**Comment by flokraft85** [comment]
Hi,
yes of cause. They look quite normal.
e4955234-38c2-425a-aef7-1d788738077f
150S2882M624I4459S
e633dac1-2ecb-4ef3-9dfd-09517808b11f
8833S4454M1799I1075S
6d3bf297-86f1-42e0-92be-2a03931fb064
7151S10969M2843I1467S
7579f2b1-8ae2-44c2-a310-a4ccc05c7c78
12S2815M630I4985S
3ce4d068-f4aa-4224-9bf1-c6099d4f3a9d
10S1457M6D22796S
9b8e95f1-8d0d-4505-aa9f-138dce20676a
14S9448M3271I3993S
2016fabb-acdd-4a0f-9811-1bd08ce22ae3
11382S10244M2411I35S

**Comment by abelhj** [comment]
I am having this exact issue as well.  It appears that the bad cigar strings (the ones with negative entries) are in the SA tags and not in column six of the bam.

```
abelhj@compute1-exec-279:/storage1/fs1/timley/Active/aml_ppg/analysis/longread/ont/analysis/sv/tumor_normal/realigned$ cat tt.txt | head -n 1 | cut -f -6,38-
b2c602bb-bfdd-46ee-bd1a-5d2f6ef274ed	256	chr1	10001	0	2785S18M1I13M1D40M1I6M1I6M1I24M2I37M4I32M1I3M1I24M1I43M1D6M1D12M2I35M3I18M3I9M1D22M2I5M1I58M25D81M1I105M7I1M1I175M3D1M3D8M7D3M1D61M44I56M2D21M1I17M1I49M2I2M1D3M1D22M1I5M2I1195M1I103M1D396M3I48M1I400M1D94M1D308M1D36M1I4M2D299M1D207M1D48M1I131M1I4M2D104M5I299M1I7M1I115M1D683M1D3M1D338M1I813M3D27M1I90M1I458M2I278M1D388M1D110M1D8M1I6M6D117M2I203M1I359M3I15M1I395M1D483M1I488M1I453M1D381M2D128M3D5M1D9M2D1M1D123M1D158M1I2M2D1M2D124M1I99M1D109M3D4M1D96M1D40M2D8M1I2M2D3M2D5M1I44M1D92M1I43M1I634M3D8M1D519M1D11M3D3M1D112M2I7M1D13M1D389M2I130M1D92M12S	rl:i:2604	SA:Z:chr1,180089,+,2244S14479M115I-16838S,16,505;chr5,10498,+,36S1384M1I-1421S,11,43;chr3,198172702,-,-2446S1772M54D674S,60,321;
abelhj@compute1-exec-279:/storage1/fs1/timley/Active/aml_ppg/analysis/longread/ont/analysis/sv/tumor_normal/realigned$ cat tt1.txt | head -n 1 | cut -f -6,38-
b2c602bb-bfdd-46ee-bd1a-5d2f6ef274ed	2048	chr5	10498	11	36H304M5I600M1I87M1D11M1D78M1D59M1I61M1D103M1D12M1D12M1D6M1D12M2I31M15429H	MD:Z:592C398^T11^T78^A120^A5A5A5A5A5A5A5A5A5A5A5A5A31^A12^G5G6^G1C4^G1C4G1C3G4C0G4C0G4C0G5G6	rl:i:2604	SA:Z:chr1,180089,+,2244H14479M115I-15453H,16,505;chr3,198172702,-,-1061H1772M54D674H,60,321;

```

**Comment by abelhj** [comment]
I am having this exact same issue.  It appears that the bad cigar strings (the ones with negative entries) are in the SA tags.

```
cat tt.txt | head -n 1 | cut -f -6,38-
b2c602bb-bfdd-46ee-bd1a-5d2f6ef274ed	256	chr1	10001	0	2785S18M1I13M1D40M1I6M1I6M1I24M2I37M4I32M1I3M1I24M1I43M1D6M1D12M2I35M3I18M3I9M1D22M2I5M1I58M25D81M1I105M7I1M1I175M3D1M3D8M7D3M1D61M44I56M2D21M1I17M1I49M2I2M1D3M1D22M1I5M2I1195M1I103M1D396M3I48M1I400M1D94M1D308M1D36M1I4M2D299M1D207M1D48M1I131M1I4M2D104M5I299M1I7M1I115M1D683M1D3M1D338M1I813M3D27M1I90M1I458M2I278M1D388M1D110M1D8M1I6M6D117M2I203M1I359M3I15M1I395M1D483M1I488M1I453M1D381M2D128M3D5M1D9M2D1M1D123M1D158M1I2M2D1M2D124M1I99M1D109M3D4M1D96M1D40M2D8M1I2M2D3M2D5M1I44M1D92M1I43M1I634M3D8M1D519M1D11M3D3M1D112M2I7M1D13M1D389M2I130M1D92M12S	rl:i:2604	SA:Z:chr1,180089,+,2244S14479M115I-16838S,16,505;chr5,10498,+,36S1384M1I-1421S,11,43;chr3,198172702,-,-2446S1772M54D674S,60,321;
cat tt1.txt | head -n 1 | cut -f -6,38-
b2c602bb-bfdd-46ee-bd1a-5d2f6ef274ed	2048	chr5	10498	11	36H304M5I600M1I87M1D11M1D78M1D59M1I61M1D103M1D12M1D12M1D6M1D12M2I31M15429H	MD:Z:592C398^T11^T78^A120^A5A5A5A5A5A5A5A5A5A5A5A5A31^A12^G5G6^G1C4^G1C4G1C3G4C0G4C0G4C0G5G6	rl:i:2604	SA:Z:chr1,180089,+,2244H14479M115I-15453H,16,505;chr3,198172702,-,-1061H1772M54D674H,60,321;


```

**Comment by tijyojwad** [comment]
thanks @flokraft85 and @abelhj for reporting - I think there's a bug in how we generate the cigar string for non-primary ones. We'll have a look!

**Comment by iiSeymour** [comment]
Fixed in [dorado v0.5.2](https://github.com/nanoporetech/dorado/releases/tag/v0.5.2).
